### PR TITLE
/status subresource and observed generation support

### LINKED ISSUE #102: Support Generation/ObservedGeneration
Per [recommendation on writing controllers](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md) Bundle object should have ObservedGeneration in it's status. Some discussions about it:
- https://github.com/kubernetes/kubernetes/issues/7328
- https://github.com/kubernetes/features/issues/95#issuecomment-302946176

It is blocked on splitting spec/status for CustomResources (item in list https://github.com/kubernetes/features/issues/95). Because Smith cannot update status without also bumping up metadata.generation (is it even incremented automatically in 1.7?).

### LINKED ISSUE #323: Make use of /status subresource
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource

This allows to update the `.spec` part without messing with `.status` and vice versa.

### PR DESCRIPTION
**Pull request by ash2k**
Fixes #323.
Fixes #102.

Need to merge https://github.com/atlassian/smith/pull/330 first.

### CODE REVIEW

File: it/utils_for_tests.go [hunk 1]
```diff
@@ -142,22 +148,14 @@ func IsBundleResourceCond(t *testing.T, namespace, name string, resource smith_v
...
 
-func IsBundleNewerCond(namespace, name string, resourceVersions ...string) watch.ConditionFunc {
+func IsBundleObservedGenerationCond(namespace, name string) watch.ConditionFunc {
 	return func(event watch.Event) (bool, error) {
 		metaObj := event.Object.(meta_v1.Object)
 		if metaObj.GetNamespace() != namespace || metaObj.GetName() != name {
 			return false, nil
 		}
 		b := event.Object.(*smith_v1.Bundle)
-		for _, rv := range resourceVersions {
```
Comment (ash2k) [inline it/utils_for_tests.go]: Die, hack, die.

File: it/utils_for_tests.go [hunk 2]
```diff
@@ -216,6 +214,18 @@ func SetupApp(t *testing.T, bundle *smith_v1.Bundle, serviceCatalog, createBundl
 	require.NoError(t, err)
 
 	defer func() {
+		cli := smithClient.SmithV1().Bundles(cfg.Namespace)
+		b, err := cli.Get(bundle.Name, meta_v1.GetOptions{})
+		if err == nil {
+			if len(b.Finalizers) > 0 {
+				t.Logf("Removing finalizers from Bundle %q", bundle.Name)
+				b.Finalizers = nil
```
Comment (ash2k) [inline it/utils_for_tests.go]: It's annoying to remove finalizers manually to make the namespace go away if the test exits abruptly.

File: pkg/controller/bundlec/bundle_sync_task.go [hunk 3]
```diff
@@ -293,42 +292,40 @@ func (st *bundleSyncTask) updateBundle() error {
...
 	if st.newFinalizers != nil {
 		// Update finalizers
 		st.bundle.Finalizers = st.newFinalizers
-		// Set status to "InProgress: True"
-		// (there will be one more iteration for resource sync that will set an appropriate status)
-		// TODO: Setting status to "InProgress" here might be unnecessary
-		// (if the Bundle spec wasn't updated during the last update, the status should be kept with no changes)
-		// The better solution could be storing "ObservedGeneration" in the status to reflect
-		// whether the status is up-to-date with the spec or stale
-		st.bundle.Status.Conditions = []cond_v1.Condition{
```
Comment (ash2k) [inline pkg/controller/bundlec/bundle_sync_task.go]: We are sending an update here to the resource's body (not `/status`) which, when subresource `/status` is enabled, ignores `.status` changes. So there is no point making any changes to `.status` - they are ignored. See https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource

File: pkg/controller/bundlec/bundle_sync_task.go [hunk 4]
```diff
@@ -293,42 +292,40 @@ func (st *bundleSyncTask) updateBundle() error {
...
-			{Type: smith_v1.BundleInProgress, Status: cond_v1.ConditionTrue},
-			{Type: smith_v1.BundleReady, Status: cond_v1.ConditionFalse},
-			{Type: smith_v1.BundleError, Status: cond_v1.ConditionFalse},
-		}
-		_, err := st.updateObjectsToDeleteStatus()
```
Comment (ash2k) [inline pkg/controller/bundlec/bundle_sync_task.go]: This is also pointless - `.status` changes are ignored now in this codepath.

File: pkg/controller/bundlec_test/finalizer_added_if_not_present_test.go [hunk 5]
```diff
@@ -87,8 +87,6 @@ func TestFinalizerAdded(t *testing.T) {
 			updateBundle := bundleUpdate.GetObject().(*smith_v1.Bundle)
 			// Make sure that the "deleteResources" finalizer was added
 			assert.True(t, resources.HasFinalizer(updateBundle, bundlec.FinalizerDeleteResources))
-
-			tc.assertObjectsToBeDeleted(t, m1)
```
Comment (ash2k) [inline pkg/controller/bundlec_test/finalizer_added_if_not_present_test.go]: When finalizer is added, we don't touch `.status` anymore (see above). So no need to test it.

File: it/utils_for_tests.go [hunk 6]
```diff
@@ -142,22 +148,14 @@ func IsBundleResourceCond(t *testing.T, namespace, name string, resource smith_v
...
-		for _, rv := range resourceVersions {
-			if b.ResourceVersion == rv {
-				// TODO Should be using Generation here once it is available
-				// https://github.com/kubernetes/kubernetes/issues/7328
-				// https://github.com/kubernetes/features/issues/95
-				return false, nil
-			}
-		}
-		return true, nil
+		return b.Status.ObservedGeneration >= b.Generation, nil
```
Comment (nilebox) [inline it/utils_for_tests.go]: Just FYI: CRDs don't get `Generation` incremented when annotations/labels updated, see https://github.com/kubernetes/kubernetes/issues/67428
I haven't checked if `ResourceVersion` get incremented in that case, but just something to keep in mind.

File: pkg/controller/bundlec/bundle_sync_task.go [hunk 7]
```diff
@@ -293,42 +292,40 @@ func (st *bundleSyncTask) updateBundle() error {
...
+		err := st.updateBundle()
```
Comment (nilebox) [inline pkg/controller/bundlec/bundle_sync_task.go]: I'm pretty sure that `status` subresource allows modifying finalizers.
I have recently found this code in Service Catalog https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_instance.go#L2420 and double checked that it actually works.
So unless CRD implementation is different, you shouldn't need to invoke this update separately from status update.
Also, there should only be one update request per processing, as we all learned the hard way.

File: it/utils_for_tests.go (hunk 6 above)
Comment (ash2k) [reply it/utils_for_tests.go]: > Just FYI: CRDs don't get Generation incremented when annotations/labels updated, see kubernetes/kubernetes#67428

Yes, and that is what I would expect. See https://github.com/atlassian/smith/issues/325 and the corresponding PR. Just found these docs https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-subresources.md#status-behavior and https://github.com/kubernetes/kubernetes/issues/45539

> I haven't checked if ResourceVersion get incremented in that case, but just something to keep in mind.

Yes, it does. It changes when a mutating request touches `/status` or `/` (normal resource path). I tested =)

File: pkg/controller/bundlec/bundle_sync_task.go (hunk 7 above)
Comment (ash2k) [reply pkg/controller/bundlec/bundle_sync_task.go]: > I'm pretty sure that status subresource allows removing finalizers.

But it should not. See the design doc and the issues I linked above.

> So unless CRD implementation is different, you shouldn't need to invoke this update separately from status update.

When adding finalizers we don't perform bundle processing and hence do not update its status.

> Also, there should only be one update request per processing, as we all learned the hard way.

Yes. That is how it was and this PR does not change this behaviour.

**Comment by ash2k** [comment]
The second commit just splits a method into two to simplify it and make the linter happy, no functional changes.

**Pull request merged**